src/rust/engine/src/lib.rs

```diff
@@ -662,6 +663,43 @@ pub extern "C" fn merge_directories(
   })
 }
 
+#[no_mangle]
+pub extern "C" fn materialize_directories(
+  scheduler_ptr: *mut Scheduler,
+  directories_paths_and_digests_value: Value,
+) -> PyResult {
+  let values = externs::project_multi(&directories_paths_and_digests_value, "dependencies");
+  let directories_paths_and_digests_results: Result<Vec<(PathBuf, Digest)>, String> = values
+    .iter()
+    .map(|value| {
+      let dir = PathBuf::from(externs::project_str(&value, "path"));
+      let dir_digest =
+        nodes::lift_digest(&externs::project_ignoring_type(&value, "directory_digest"));
+      dir_digest.map(|dir_digest| (dir, dir_digest))
+    })
+    .collect();
+
+  let dir_and_digests = match directories_paths_and_digests_results {
+    Ok(d) => d,
+    Err(err) => {
+      let e: Result<Value, String> = Err(err);
+      return e.into();
+    }
+  };
+
+  with_scheduler(scheduler_ptr, |scheduler| {
+    let store = scheduler.core.store.clone();
```

Comment:
I think you can `&scheduler.core.store` rather than `scheduler.core.store.clone()` here - you don't actually need to move `store` anywhere, so a reference should work fine :)

Reply:
It didn't work passing a reference, but I also didn't need to clone the store. Changes to come.